packages/main/src/index.ts

```diff
@@ -44,34 +45,65 @@ const createWindow = async () => {
   mainWindow.on('ready-to-show', () => {
     mainWindow?.show();
 
-    if (import.meta.env.MODE === 'development') {
+    if (isDevelopment) {
       mainWindow?.webContents.openDevTools();
     }
   });
 
-  /**
-   * External hyperlinks open in the default browser.
-   *
-   * @see https://stackoverflow.com/a/67409223
-   */
-   mainWindow.webContents.setWindowOpenHandler(({ url }) => {
-    shell.openExternal(url);
-    return { action: 'deny' };
-  });
-
   /**
    * URL for main window.
    * Vite dev server for development.
    * `file://../renderer/index.html` for production and test
    */
-  const pageUrl = import.meta.env.MODE === 'development' && import.meta.env.VITE_DEV_SERVER_URL !== undefined
+  const pageUrl = isDevelopment && import.meta.env.VITE_DEV_SERVER_URL !== undefined
     ? import.meta.env.VITE_DEV_SERVER_URL
     : new URL('../renderer/dist/index.html', 'file://' + __dirname).toString();
 
 
   await mainWindow.loadURL(pageUrl);
 };
 
+ app.on('web-contents-created', (_event, contents) => {
+
+  /**
+   * Empty since there's no reason to enable any navigation.
+   *
+   * @see https://www.electronjs.org/docs/latest/tutorial/security#13-disable-or-limit-navigation
+   */
+  contents.on('will-navigate', (event, url) => {
+    const trustedOrigins = new Set();
+    const trustedProtocols = new Set();
+    const { origin, protocol } = new URL(url);
+    if (!trustedOrigins.has(origin) || !trustedProtocols.has(protocol)){
+      console.warn('Blocked navigating to an unrecognized origin:', origin);
+      event.preventDefault();
+    }
+  });
+
+  /**
+  * Hyperlinks to trusted sites open in the default browser.
+  *
+  * @see https://www.electronjs.org/docs/latest/tutorial/security#14-disable-or-limit-creation-of-new-windows
+  * @see https://www.electronjs.org/docs/latest/tutorial/security#15-do-not-use-openexternal-with-untrusted-content
+  */
+  contents.setWindowOpenHandler(({ url }) => {
+    const trustedOrigins = new Set([
+      'https://vitejs.dev',
+      'https://github.com',
+      'https://v3.vuejs.org']);
+    const trustedProtocols = new Set([
+      'https:', // https://www.electronjs.org/docs/latest/tutorial/security#1-only-load-secure-content
+    ]);
+    const { origin, protocol } = new URL(url);
+    if (trustedOrigins.has(origin) && trustedProtocols.has(protocol)){
```

Comment:
I think this check is a bit excessive. There is no need to check the protocol, because the original already contains the protocol.

Reply:
I agree there's no _need_ for the protocol check. However, I think that it's good to have explicit logic testing the protocol to ensure that people don't add something like `http://google.com`. I can imagine a scenario where someone runs into this check, grumbles, and then _adds_ `https://google.com`, leaving `http://google.com`. Yes, the protocol is in the origin, but programmers are a silly lazy bunch so I thought I'd make it hard to screw up.

Side note, I'm traveling a bit over the next few days, so I'll be slow to respond/make changes to this PR.

Reply:
I think excessive checking of the protocol is superfluous. There's nothing stopping a lazy programmer from adding `http:` to trusted protocols. Therefore, it does not provide significant protection.

However, I agree that a little hint is needed regarding safety requirements. I think it will be enough to specify it at the level of types:
```ts
const trustedOrigins = new Set<`https:${string}`>()
```

It still does not provide a significant level of protection against a careless developer and can be extended as needed, but such a record is shorter, clearer and exists only in compile time.

Reply:
Oh neat - I didn't know template literals could be used in generics. I'm very new to Typescript (and webdev in general).